### ISSUE TITLE
Some rules unnecessarily slow because of unneeded creation of import map

### ISSUE THREAD

**Issue by jorisw** [import-js/eslint-plugin-import#2953]
Hi,

While investigating the performance of our linter setup, I ran a `TIMING=1` type performance scan of my linter rules, and found that unfortunately, when any of the used rules needs an import map, it will look like the first rule who uses it, takes 90+% of the work. Example:

```
> DEBUG=eslint:cli-engine TIMING=1 yarn lint --debug --verbose

All files pass linting.

Rule                              | Time (ms) | Relative
:---------------------------------|----------:|--------:
import/namespace                  | 10247.586 |    98.0%
import/no-cycle                   |   114.264 |     1.1%
@nx/enforce-module-boundaries     |    30.531 |     0.3%
import/no-unresolved              |    10.439 |     0.1%
@typescript-eslint/no-unused-vars |     6.393 |     0.1%
```

`eslint-plugin-import` seems to suffer from this most, in fact, when I disable all `import/*` rules, the above issue goes away: the time spent between rules is more balanced. Example:

```
Rule                              | Time (ms) | Relative
:---------------------------------|----------:|--------:
@nx/enforce-module-boundaries     |    16.948 |    22.5%
@typescript-eslint/no-unused-vars |    13.912 |    18.4%
jest/no-duplicate-hooks           |     2.725 |     3.6%
sonarjs/cognitive-complexity      |     1.869 |     2.5%
sonarjs/no-gratuitous-expressions |     1.668 |     2.2%
```

Discussing with the ESLint maintainers how I might overcome the misbalance in reported time spent, one notable statement came back regarding `eslint-plugin-import`: 

> Overall, `eslint-plugin-import` rules always try to resolve every import even when they don't need to (looking at you, `no-unused-imports`), and that is, in fact, what makes them slow. The maintainers seem to think that other plugins/rules will prepopulate the import map for them, but that's just not the case in the current ecosystem.

Would you say this is accurate? Could the the rules who don't need the import map, be made not to trigger the building of one? Or the other way around? Granted I don't fully understand the implication, or its implications.

**Comment by soryy708** [comment]
How do we move forward with this?

We build a full ExportMap only in rules that use the ExportMap.
Are there any rules that use ExportMap that can be optimized to avoid it?

**Comment by TkDodo** [comment]
I think I have similar observations working on the [getsentry/sentry](https://github.com/getsentry/sentry) codebase. `import/no-duplicates` and `import/no-nodejs-modules` are the ones that show up as needing a lot of time for me:

```
Rule                                                       | Time (ms) | Relative
:----------------------------------------------------------|----------:|--------:
import/no-duplicates                                       |  8646.202 |    16.7%
@typescript-eslint/naming-convention                       |  7779.880 |    15.0%
@typescript-eslint/no-unnecessary-type-assertion           |  3605.909 |     6.9%
import/no-nodejs-modules                                   |  3149.472 |     6.1%
```

When I turn off `import/no-duplicates`, `import/no-nodejs-modules` takes more time. When I turn them both off, the total time drops significantly, by around 13% on my machine.

We use more rules from `eslint-plugin-import`, so I’m really not sure, but my guess is those are the only two we have that need those import maps?

**Comment by TkDodo** [comment]
@JounQin I understand that to some degree, but we use other rules from `eslint-plugin-import` that are apparently not “affected” by this “parsing”, because they are always fast. For example, `import/no-webpack-loader-syntax` takes 44ms in all cases.

So I’m wondering what kind of extra parsing e.g. `import/no-nodejs-modules` needs ? Shouldn’t both rules (`import/no-nodejs-modules` and `import/no-webpack-loader-syntax`) just look at each import statement in isolation and determine if it’s valid or not?

**Comment by ljharb** [comment]
Indeed, the only rules that can show up in this "slowness" are the ones that *do* need the ExportMap created.

If eslint provided us with a "pre lint" phase, we'd create the map there, and the timing results would be much more consistent for individual rules.
